xarray/coding/variables.py

```diff
@@ -232,20 +297,46 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
     return data
 
 
-def _choose_float_dtype(dtype: np.dtype, has_offset: bool) -> type[np.floating[Any]]:
-    """Return a float dtype that can losslessly represent `dtype` values."""
-    # Keep float32 as-is.  Upcast half-precision to single-precision,
+def _choose_float_dtype(
+    dtype: np.dtype, mapping: MutableMapping
+) -> type[np.floating[Any]]:
+    # check scale/offset first to derive wanted float dtype
+    # see https://github.com/pydata/xarray/issues/5597#issuecomment-879561954
+    scale_factor = mapping.get("scale_factor")
+    add_offset = mapping.get("add_offset")
+    if scale_factor or add_offset:
+        # get the type from scale_factor/add_offset to determine
+        # the needed floating point type
+        if scale_factor:
+            scale_type = type(scale_factor)
+        if add_offset:
+            offset_type = type(add_offset)
+        # CF conforming, both scale_factor and add-offset are given and
+        # of same floating point type
+        if (
+            add_offset
+            and scale_factor
+            and offset_type == scale_type
+            and np.issubdtype(scale_type, np.floating)
+        ):
+            return np.dtype(scale_type)
```

Comment:
```suggestion
            return np.dtype(scale_type).type
```
Probably want the same return type as the others.

```python
np.dtype(np.float32)
Out[6]: dtype('float32')

np.dtype(np.float32).type
Out[7]: numpy.float32

# Can't call dtypes:
np.dtype(np.float32)(1)
TypeError: 'numpy.dtype[float32]' object is not callable

np.dtype(np.float32).type(1)
Out[13]: 1.0
```

Reply:
Oh, my bad... Thanks @Illviljan, your help is much appreciated. BTW, the new batch committing in the GitHub UI is very comfortable.